### PR TITLE
[DNM]kv: XAPI support basic select.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: kv/kv.go [hunk 1]
```diff
@@ -74,6 +79,28 @@ type Transaction interface {
...
+
+	// StartTS returns the start timestamp.
+	StartTS() int64
+}
+
+// XAPI represents extended KV API, it is more efficient than basic KV API.
+type XAPI interface {
+	// Select sends a select request to kv store.
+	// If concurrent is set to 1, the result is returned in order.
+	Select(req *tipb.SelectRequest, concurrency int) SelectResult
```
Comment (shenli) [inline kv/kv.go]: Will there be more functions other than Select?

File: executor/builder.go [hunk 2]
```diff
@@ -117,6 +117,25 @@ func (b *executorBuilder) buildFilter(src Executor, conditions []ast.ExprNode) E
...
+		return nil
+	}
+	xapi := txn.GetXAPI()
+	if xapi != nil {
+		e := &XSelectTableExec{
+			ctx:       b.ctx,
+			tablePlan: v,
+		}
+		where := conditionsToPBExpression(v.FilterConditions)
+		if xapi.SupportExpression(where) {
```
Comment (shenli) [inline executor/builder.go]: Can we check SupportExpression() before check if xapi != nil ?

Comment (coocood) [inline executor/builder.go]: Even if expression is not supported, we can still use xapi, we just filter the result in tidb-server side.

File: kv/kv.go (hunk 1 above)
Comment (coocood) [inline kv/kv.go]: Maybe we can support insert/delete/update many rows.

File: kv/kv.go [hunk 3]
```diff
@@ -13,6 +13,11 @@
 
 package kv
 
+import (
+	"github.com/pingcap/tidb/kv/tipb"
+	"github.com/pingcap/tidb/util/types"
```
Comment (shenli) [inline kv/kv.go]: types is not used in this file?

File: kv/kv.go (hunk 1 above)
Comment (nieyy) [inline kv/kv.go]: How about `concurrency` to `parallel`?

File: kv/kv.go [hunk 4]
```diff
@@ -74,6 +79,28 @@ type Transaction interface {
 	SetOption(opt Option, val interface{})
 	// DelOption deletes an option.
 	DelOption(opt Option)
+
+	// GetXAPI gets XAPI interface if it is supported.
+	// Otherwise returns nil.
+	GetXAPI() XAPI
+
+	// StartTS returns the start timestamp.
+	StartTS() int64
```
Comment (nieyy) [inline kv/kv.go]: `StartTS` -> `GetStartTimestamp`?

Comment (ngaut) [inline kv/kv.go]: StartTS is neat.

File: kv/kv.go [hunk 5]
```diff
@@ -74,6 +79,28 @@ type Transaction interface {
...
+type XAPI interface {
+	// Select sends a select request to kv store.
+	// If concurrent is set to 1, the result is returned in order.
+	Select(req *tipb.SelectRequest, concurrency int) SelectResult
+	// SupportExpression checks if the expression is supported.
+	SupportExpression(expr *tipb.Expression) bool
+}
+
+// SelectResult is used to get response rows from SelectRequest.
+type SelectResult interface {
```
Comment (nieyy) [inline kv/kv.go]: `SelectResult` -> `SelectIterator`?

File: kv/kv.go (hunk 1 above)
Comment (coocood) [inline kv/kv.go]: In Go, `parellel` is a rarely used term.

File: kv/kv.go [hunk 6]
```diff
@@ -74,6 +74,48 @@ type Transaction interface {
...
+// Request represents a kv request.
+type Request struct {
+	// The request type.
+	Tp       int64
+	Data     []byte
+	StartKey []byte
+	EndKey   []byte
+	// If desc is true, the request is sent in descending order.
+	Desc bool
+	// If concurrency is greater than 1, it concurrently request multiple region servers.
```
Comment (nieyy) [inline kv/kv.go]: If `concurrency` is greater than 1, the request will be sent to multiple storage nodes concurrently.

The noun `region server` is specific to HBase or TiKV only.

Comment (coocood) [inline kv/kv.go]: OK

File: kv/kv.go [hunk 7]
```diff
@@ -74,6 +74,48 @@ type Transaction interface {
...
+	Concurrency int
+}
+
+// ResponseIterator is used to get responses from multiple region.
```
Comment (nieyy) [inline kv/kv.go]: ResponseIterator is used to get responses from KV layer.

`region` is for HBase or TiKV only.

Comment (ngaut) [inline kv/kv.go]: No, as we discussed last time, we will use the concepts of HBase.

**Comment by shenli** [comment]
LGTM

**Comment by nieyy** [comment]
LGTM

**Pull request closed without merging**